p3387.tex

```diff
@@ -154,25 +160,25 @@ \subsection{Point of evaluation of \tcode{pre} and \tcode{post}}
 
 \cite{P2900R8} specifies that the precondition assertions are evaluated ``immediately after function parameters are initialised and before entering the function body''. For a coroutine, by ``function body'', we do \emph{not} mean the coroutine body that the user wrote, but the body of the ramp function that the compiler generated. Therefore, the precondition assertions of a coroutine are evaluated \emph{before} any of the coroutine-specific events that happen in the ramp function, such as creating a copy of the parameters, allocating the coroutine state, initialising the promise object, etc.
 
-\cite{P2900R8} further specifies that the postcondition assertions are evaluated ``after the return value has been initialised and local automatic variables have been destroyed but prior to the destruction of function parameters''. Again, for a coroutine, this specification should be applied to the body of the ramp function and not the coroutine body. This means that the ``return value'' is that of the ramp function, \emph{not} the values yielded by the coroutine, and its type is the declared return type of the coroutine. Likewise, the ``local variables'' are those created by the ramp function, \emph{not} the local variables in the coroutine body.
+\cite{P2900R8} further specifies that the postcondition assertions are evaluated ``after the return value has been initialised and local automatic variables have been destroyed but prior to the destruction of function parameters''. Again, for a coroutine, this specification should be applied to the body of the ramp function and not the coroutine body. This means that the ``return value'' is that of the ramp function, \emph{not} (for example) values yielded by the coroutine; its type is the declared return type of the coroutine. Likewise, the ``local variables'' are those created by the ramp function, \emph{not} the local variables in the coroutine body.
```

Comment:
I think that the important distinction is that we are destroying local variables that would be destroyed by returning --- returning to the caller from a coroutine (through a co_await or something equivalent to a co_await) involves leaving the scope of the coroutine *without* destroying local parameters.   Even in such cases, we must still have initialized the return value and we still destroy the function parametsr on a call.

Reply:
perhaps we should say "are those created _for_ the ramp function, by the compiler's implementation"

Reply:
we could... but is there anything observable that is destroyed when a coroutine suspends the first time?   Function parameters are, but their lifetimes are governed by the function call, and not the function - that's all just specified in expr.call, and independent of anything coroutine-related.

any truly local variables to the ramp function are inivisble things as far as any user is concerned.   In truth, can't think of what would have a non-trivial destructor that would be invoked when you return for the first time.

Reply:
works for me

Reply:
agreed all the local state of the ramp _is_ invisible and un-inspectable to the user - I was just trying to clarify that the local state was not that of the coroutine body.  I'm fine with whatever you folks think makes that clear.

Reply:
I now phrased it as follows, hope that's OK:
"This means that the ``return value'' is that of the ramp function, \emph{not} (for example) values yielded by the coroutine; its type is the declared return type of the coroutine; and there are no ``local variables'' since the local variables of the coroutine body are not in the scope of the ramp function and are not destroyed when the latter returns."